homeassistant/components/hassmpris/__init__.py

```diff
@@ -0,0 +1,102 @@
+"""The MPRIS media playback remote control integration."""
+from __future__ import annotations
+
+from typing import Any, cast
+
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey
+import hassmpris_client
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+
+from .cert_data import CertStore
+from .const import CONF_HOST, CONF_MPRIS_PORT, DOMAIN, LOGGER as _LOGGER
+from .models import ConfigEntryData, HassmprisData
+
+PLATFORMS: list[Platform] = [Platform.MEDIA_PLAYER]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up MPRIS media playback remote control from a config entry."""
+
+    # This would be an error, and the type checker knows it.
+    assert entry.unique_id is not None
+
+    try:
+        client_cert, client_key, trust_chain = await CertStore(
+            hass, entry.unique_id
+        ).load_cert_data()
+    except KeyError as exc:
+        raise ConfigEntryAuthFailed(
+            "Authentication data is missing -- must reauth"
+        ) from exc
+
+    entry_data = cast(ConfigEntryData, entry.data)
+    clnt = hassmpris_client.AsyncMPRISClient(
+        entry_data[CONF_HOST],
+        entry_data[CONF_MPRIS_PORT],
+        client_cert,
+        cast(RSAPrivateKey, client_key),
+        trust_chain,
+    )
+    try:
+        _LOGGER.debug("Pinging the MPRIS agent")
+        await clnt.ping()
+        _LOGGER.info("Successfully pinged the MPRIS agent")
+
+    except hassmpris_client.Unauthenticated as exc:
+        raise ConfigEntryAuthFailed(exc) from exc
+    except Exception as exc:
+        raise ConfigEntryNotReady(str(exc)) from exc
+
+    hass.data.setdefault(DOMAIN, {})
+    component_data = hass.data[DOMAIN][entry.entry_id] = HassmprisData(
+        clnt, [], None, None
+    )
+
+    async def _run_unloaders(*unused_args: Any) -> None:
+        _LOGGER.debug("Running unloaders")
+        while component_data.unloaders:
+            await component_data.unloaders.pop()()
+        _LOGGER.debug("Unloaders complete")
+
+    component_data.unload_func = _run_unloaders
```

Comment:
This is just unnecessary. Move `_run_unloaders` to module scope instead so it can be called from `async_unload_entry`, and make it take a `HassmprisData` instance as its only argument.

Reply:
Very much nicer code as a result.  Thanks.